hypothetical carbon stocks for emission pricing

flohump opened the pull request:
## :bird: Purpose of this PR :bird:

- PR386 partly solved one problem under CO2 emission pricing (decline in primary forest) but introduced another problem (large shift from cropland and pasture to other land)
- This PR aims to solve both problems by introducing hypothetical carbon stocks for emission pricing
 - **modules** Moved interface `vm_carbon_stock` from 52_carbon to 56_ghg_policy
 - **modules** New dimension in `vm_carbon_stock` for different carbon stock types (actual, previousLandPattern, previousCarbonDensity)
 - **config** added option for CO2 emission pricing `cfg$gms$c56_carbon_stock_pricing`
 - **config** added cfg$gms$s70_past_mngmnt_factor_fix with default 2005 (previous default was 2010). The previous setting caused a strong spike in CO2 emissions from pasture expansion in SSA. With 2005, this can be avoided.

 - **config** cfg$gms$s13_max_gdp_shr changed back to 1 as default because 0.002 caused problems in some cases.

- **56_ghg_policy** Deactivated GHG emission policies were not accounted for in the MACCs module. This has been corrected by an extension of the interface `im_pollutant_prices`, which now has an additional dimension for emission sources `emis_source`. In this context some equations in `56_ghg_policy` have been simplified (sets: `emis_source_reg`, `emis_source_cell`). Also, GHG emissions from peatlands have been fully integrated into `56_ghg_policy`.

- I made the following tests. In these SSP2-based runs, there should be no loss of primary forest from 2050 onwards.

#option 0: Before PR386. Version with declining primary forest. Showing the original problem.
  c56_carbon_stock_pricing <- "actual"
  c56_emis_policy <- "redd+natveg_nosoil"
  s56_reward_neg_emis <- 0 

#option 00: Current develop (introduced with PR386). In this setup the primary forest issue is (partly) solved but there is strong conversion of cropland/pasture to other land.
  c56_carbon_stock_pricing <- "actual"
  c56_emis_policy <- "redd+natveg_nosoil"
  s56_reward_neg_emis <- -Inf

#option 1: minimal change. Instead of actual carbon stocks, CO2 emissions used for emission pricing are based on carbon stocks that depend on current area and previous carbon density. Thus, effectively only emissions from land-use change are priced, which should solve the primary forest issue.
  c56_carbon_stock_pricing <- "previousCarbonDensity"
  c56_emis_policy <- "redd+natveg_nosoil"
  s56_reward_neg_emis <- 0

#option 2: medium change. Negative emissions in existing areas are rewarded. This should solve the primary forest issue. Negative emissions from newly established areas are excluded from the emissions pricing (Content-wise this is covered by the afforestation implementation). This avoids strong conversion of cropland/pasture to other land. 
  c56_carbon_stock_pricing <- "actualNoAcEst"
  c56_emis_policy <- "redd+natveg_nosoil"
  s56_reward_neg_emis <- -Inf

Results: 
Primary forest and afforestation:
Option 0 is the problematic case. This is only shown as a point of reference. Decline of primary forest and increase of afforestation after 2050.
Option 00 is also problematic. Strong conversion of cropland/pasture to other land
Option 1 partly solves the primary forest issue, but not fully (2070->2080). In addition, there is more loss before 2025, i.e. before the carbon price starts, compared to option 2. It seems that primary forest is partly used for afforestation between 2070 and 2090.
Option 2 solves the primary forest issue and generates reasonable afforestation patterns. 

The difference in primforest dynamics is also visible in CO2 emissions until 2025.

Other variables like tau, cropland, pasture and other land do not differ much between options 0, 1 and 2.

Considering these outcomes, I suggest to go for option 2. 
For option 2, I also made tests with c56_emis_policy <- "all_nosoil", which produced reasonable results. 
If we agree on option 2, we should consider removing s56_reward_neg_emis.

## :wrench: Checklist for PR creator :wrench:

- [x] Labeling pull request correctly [from the label list](https://github.com/magpiemodel/magpie/labels).

* Low risk : Simple bugfixes (missing files, updated documentation, typos) or Start/output scripts
* Medium risk : New realization / Changes to existing realization / Other changes which don't modify default.cfg
* High risk : New input files (if cfg$input is changed in default.cfg) / Modification to core model (eg. changes in equations, calculations, introduction of new sets etc.) / Other changes in default.cfg

- [x] Providing additional information based on PR label

* Low risk : No new model run needed.
* Medium risk : Default run based on the current version of the fork from which PR is made
* High risk
  * Default run from the current develop branch: PR_CO2_before_SSP2-Ref
  * Default run based on the current version of the fork from which PR is made: PR_CO2_after11_SSP2-Ref

**Default runs**
 * CO2 emissions before 2020 are different because of cfg$gms$s70_past_mngmnt_factor_fix with default 2005 (previous default was 2010)
 
![Emissions_CO2_Land_Land_use_Change-49](https://user-images.githubusercontent.com/16921122/165473592-4efbdf15-0987-4135-8369-822a1a12712c.png)
![Resources_Land_Cover_Change_Forest_Natural_Forest_Primary_Forest-11](https://user-images.githubusercontent.com/16921122/165473596-aa6ff7e7-7b04-4720-8ae5-d382f168e1b5.png)
![Resources_Land_Cover_Change_Other_Land-6](https://user-images.githubusercontent.com/16921122/165473603-5fbd5aa3-4cbe-418d-9b9e-e9606385c0e2.png)
![Resources_Land_Cover_Cropland-67](https://user-images.githubusercontent.com/16921122/165473608-bce009d6-436c-481c-93d8-9ec821efe681.png)
![Resources_Land_Cover_Forest_Managed_Forest-8](https://user-images.githubusercontent.com/16921122/165473612-dbf833e0-08d3-421c-8468-caaddf9a6bf0.png)
![Resources_Land_Cover_Pastures_and_Rangelands-29](https://user-images.githubusercontent.com/16921122/165473614-01312329-70c1-4456-ad21-71b9e20d81e5.png)


**Runs with CO2 pricing:** 

![Resources_Land_Cover_Change_Forest_Natural_Forest_Primary_Forest-12](https://user-images.githubusercontent.com/16921122/165474204-f05984c0-0e2d-40cf-a425-bc426ecfd3af.png)
![Resources_Land_Cover_Change_Other_Land-7](https://user-images.githubusercontent.com/16921122/165474216-795de1bf-d960-41a1-be01-79c28a3ff0b4.png)
![Resources_Land_Cover_Forest_Managed_Forest_Afforestation-9](https://user-images.githubusercontent.com/16921122/165474219-f38cdb2e-8223-4442-95de-4f2adc727d58.png)
![Resources_Land_Cover_Pastures_and_Rangelands-30](https://user-images.githubusercontent.com/16921122/165474366-40bd19c7-18b9-44e0-a177-7958e865f33f.png)


- [x] :chart_with_downwards_trend: Performance loss/gain from current default behavior :chart_with_upwards_trend:
  * Current develop branch default : 0.48 h
  * This PR's default :  0.44 h

* Runtime is only marginally affected by this PR.

![Unknown-72](https://user-images.githubusercontent.com/16921122/165474475-dccf6f26-e1d5-455f-a09c-5b2aad443a51.png)

- [x] Added changes to `CHANGELOG.md`
- [x] Compilation check (model starts without compilation errors - use `gams main.gms action=c` in model folder for testing).
- [x] No hard coded numbers and cluster/country/region names.
- [x] The new code doesn't contain declared but unused parameters or variables.
- [x] Where relevant, In-code comments added including documentation comments.
- [x] Made sure that documentation created with [`goxygen`](https://github.com/pik-piam/goxygen) is okay (use `goxygen::goxygen()` for testing).
- [x] Changes to [`magpie4`](https://github.com/pik-piam/magpie4) R library for post processing of model output (ideally backward compatible).
- [x] Self-review of my own code.
- [x] For high risk runs: validation of major model indicators - Land-use, emissions, food prices, Tau.  %Delete this line in case it is not a high risk run%

## :warning: Additional notes or warnings :warning:
NA

## :rotating_light: Checklist for RSE reviewer :rotating_light:

- [x] PR is labeled correctly.
- [x] `CHANGELOG` is updated correctly
- [x] No hard coded numbers and cluster/country/region names.
- [x] No unnecessary increase in module interfaces
- [x] All required updates in interfaces (if any) have been properly adressed in the module contracts
- [x] In-code comments and documentation comments are satisfactory.
- [x] model behavior/performance is satisfactory.
- [x] Requested changes (if any) were applied correctly

## :rotating_light: Checklist for MAgPIE reviewer :rotating_light:

- [ ] PR is labeled correctly.
- [ ] `CHANGELOG` is updated correctly
- [ ] No hard coded numbers and cluster/country/region names.
- [ ] Changes to the model are scientifically sound
- [ ] In-code comments and documentation comments are satisfactory.
- [ ] model behavior/performance is satisfactory.
- [ ] Requested changes (if any) were applied correctly

tscheypidi commented on config/default.cfg:
```diff
@@ -1194,15 +1197,18 @@ cfg$gms$c56_emis_policy <- "redd+natveg_nosoil" 		# def = redd+natveg_nosoil
-# * Note 2: 	s56_reward_neg_emis = -Inf in combination with pricing CO2 emissions from soils
-# * 		(see c56_emis_policy) provides a strong incentive for conversion of cropland
-# * 		to other land, given the considerably lower soil carbon content of cropland.
+# * Please not the choice of s56_reward_neg_emis strongly interacts with the choice of c56_carbon_stock_pricing.
```
```suggestion
# * Please note the choice of s56_reward_neg_emis strongly interacts with the choice of c56_carbon_stock_pricing.
```

flohump replied:
updated.

tscheypidi commented on core/calculations.gms:
```diff
@@ -32,15 +32,18 @@ file dummy; dummy.pw=2000; put dummy;
+	  
       display "Year";
       display ct;
+      display pt;
```
please remove this display statement or provide some context as it is done for ct. Otherwise the output will be confusing

flohump replied:
done.

tscheypidi commented on modules/31_past/endo_jun13/declarations.gms:
```diff
@@ -8,7 +8,10 @@
 equations
  q31_prod(j)              Cellular pasture production constraint (mio. tDM per yr)
- q31_carbon(j,ag_pools)   Above ground carbon content calculation for pasture (mio tC)
+ q31_carbon1(j,ag_pools)  Above ground carbon content calculation for pasture (mio tC)
```
I thought it should be possible to write `q31_carbon(j, ag_pools, stockType)` instead and then declare this equation separately for each element  of stockType. Have you tested this? If this works, it would improve readability

flohump replied:
Done.

tscheypidi commented on modules/31_past/endo_jun13/equations.gms:
```diff
@@ -19,10 +19,22 @@ q31_prod(j2) ..
+         sum(ct, pcm_land(j2,"past")*fm_carbon_density(ct,j2,"past",ag_pools));
+
+q31_carbon4(j2,ag_pools) ..
+ vm_carbon_stock(j2,"past",ag_pools,"previousCarbonDensity") =e=
```
alternatively to my suggestion above one could also write everything in one equation and use dollar conditions to distinguish the 4 different cases

flohump replied:
I addded 2 macros for the carbon stock calculation.

tscheypidi commented on modules/32_forestry/dynamic_feb21/equations.gms:
```diff
@@ -86,10 +86,22 @@ sum(ac_est, v32_land(j2,"aff",ac_est)) =l= sum(ac, v32_land(j2,"aff",ac)) - sum(
+ q32_carbon4(j2,ag_pools)  .. vm_carbon_stock(j2,"forestry",ag_pools,"previousCarbonDensity") =e=
+                         sum((type32,ac), v32_land(j2,type32,ac)*
+                         sum(pt, p32_carbon_density_ac(pt,j2,type32,ac,ag_pools)));
+
```
would it make sense to work with a macro here? the formulations look all pretty similar across the different land modules

flohump replied:
Yes. I addded 2 macros for the carbon stock calculation.

tscheypidi reviewed (changes requested):
Looks good overall. A question and an update request:

* You changed some interfaces. Have you checked that the outputs scripts and the functions in `magpie4` are all compatible to these changes (functions in `magpie4` in addition need to remain downwards compatible)

* The code is currently quite repetivite. I would suggest the following: Merge the carbon equations 1,2,3,4 again into one equation using dollar conditions and write a macro which produces this equation for the different land modules.

tscheypidi commented:
another question: Did you create the goxygen documentation and check whether the module descriptions now fit the new structure/interfaces of the modules? In the code I saw some documentation updates but overall had the feeling that some additional updates might be required here. In particular, the descriptions in the various `module.gms` files might now be outdated

tscheypidi commented:
and another question (sorry): can you show runtime comparison using the same GAMS version. I remember Marcos mentioning that runs are now around 40% faster. If this is the case this PR seems to slow down the model again quite significantly. Would be important to know what the impact of these changes on performance is

flohump commented:
> another question: Did you create the goxygen documentation and check whether the module descriptions now fit the new structure/interfaces of the modules? In the code I saw some documentation updates but overall had the feeling that some additional updates might be required here. In particular, the descriptions in the various `module.gms` files might now be outdated

I made some additional revisions in the documentation, espcially in 56_ghg_policy.

flohump commented:
> and another question (sorry): can you show runtime comparison using the same GAMS version. I remember Marcos mentioning that runs are now around 40% faster. If this is the case this PR seems to slow down the model again quite significantly. Would be important to know what the impact of these changes on performance is

This PR has only minimal effects on run time. See the additinal figure for run time.

flohump commented:
> Looks good overall. A question and an update request:
> 
> * You changed some interfaces. Have you checked that the outputs scripts and the functions in `magpie4` are all compatible to these changes (functions in `magpie4` in addition need to remain downwards compatible)

Yes, I updated the magpie4 carbonstock function (downwards compatible).
https://github.com/pik-piam/magpie4/commit/0e9e90a33409690bdee6a3680a510f2476291e58
The test runs shown here used the updated version. 

> * The code is currently quite repetivite. I would suggest the following: Merge the carbon equations 1,2,3,4 again into one equation using dollar conditions and write a macro which produces this equation for the different land modules.

I added 2 macros for carbon stock calcuatio, which allows to considerably reduced the number of required equaitons.

tscheypidi reviewed (approved):
looks good

tscheypidi commented on config/default.cfg:
```diff
@@ -1209,11 +1209,6 @@ cfg$gms$s56_reward_neg_emis <- -Inf			# def = -Inf
-# * Peatland GHG emission pricing
-# * 0=off
-# * 1=on
-cfg$gms$s56_peatland_policy <- 1			# def = 1
```
so this is not needed anymore?

flohump replied:
Yes. This is now covered by `c56_emis_policy`.

tscheypidi commented on core/sets.gms:
```diff
@@ -300,19 +300,19 @@ sets
    / inorg_fert, man_crop, awms, resid, man_past, som,
-     rice, ent_ferm, beccs /
+     rice, ent_ferm, resid_burn /
 
```
was `beccs` in there before by mistake or why is it now taken out / replaced by `peatland` and `resid_burn`?

flohump replied:
`beccs` was unused and therefore removed.
`resid_burn` was missing.

tscheypidi commented on modules/35_natveg/dynamic_feb21/scaling.gms:
```diff
@@ -5,8 +5,9 @@
 v35_other_expansion.scale(j,ac) = 10e-3;
 v35_other_reduction.scale(j,ac) = 10e-3;
+v35_secdforest_expansion.scale(j,ac) = 10e-3;
+v35_secdforest_reduction.scale(j,ac) = 10e-3;
```
what is the purpose of these various scaling updates?

flohump replied:
I removed some variables and changed the dimension of other variables. 
Therefore, I used gdx::calc_scaling("fulldata.gdx") for `SSP-Ref` and `SSP-Pk900` to get updated scaling factors. I also noticed that some scaling factors had been declared twice, in the cost module and the module defining the varible. In that case, I kept the later one.

tscheypidi commented on modules/51_nitrogen/ipcc2006_sep16/presolve.gms:
```diff
@@ -5,8 +5,9 @@
-
- vm_btm_reg.fx(i,emis_source_nonitrogen51,n_pollutants) = 0;
+vm_btm_reg.lo(i,emis_source_n51,n_pollutants) = -Inf;
+*vm_btm_reg.lo(i,"som",n_pollutants) = -Inf;
```
please don't add commented code
```suggestion
```

tscheypidi commented on modules/56_ghg_policy/price_jan20/preloop.gms:
```diff
@@ -9,15 +9,12 @@
+vm_btm_cell.fx(j,emis_source_cell,pollutants) = 0;
+vm_btm_cell.lo(j,emis_co2,"co2_c") = -Inf;
+vm_btm_cell.up(j,emis_co2,"co2_c") = Inf;
+***fix to zero be default here. Bounds are released in the respective modules.
```
why is that fixed to 0?

flohump replied:
To avoid unbounded variable dimensions.

tscheypidi replied:
i think that should be taken care off by the modules filling in the information.

flohump replied:
I moved the fixing of variables to the responsible modules.

tscheypidi commented on modules/58_peatland/on/equations.gms:
```diff
@@ -105,6 +105,14 @@
+
+ q58_peatland_emis_n2o(j2) ..
+	vm_btm_cell(j2,"peatland","n2o_n_direct") =e= v58_peatland_emis(j2,"n2o")/298*28/44;
+	
```
please improve the syntax:

* use a single equation instead of three separate ones
* store the conversion factors in a parameter
* create a mapping between emission types and sum over this mapping

tscheypidi reviewed (changes requested):
I put some requests for technical imrpovements. I was focussing on technical aspects, the contenwise changes need to reviewed by someone else.

bodirsky commented on modules/59_som/cellpool_aug16/equations.gms:
```diff
@@ -58,9 +58,16 @@ q59_som_pool_noncropland(j2, noncropland59) ..
 *' and the reference soil carbon pool of the subsoil
 
 q59_carbon_soil(j2,pools59) ..
```
are always all three calculations done, or are they alternatively selected? Why do you need all three in parallel?

bodirsky reviewed (changes requested):
a few open questions. Some may need discussion, as I was not always sure whether they need change.

flohump commented on config/default.cfg:
```diff
@@ -275,7 +275,7 @@ cfg$gms$c13_tccost <- "medium"		# def = medium
 
 # * Maximum regional tech cost expressed as share of reginal GDP
-cfg$gms$s13_max_gdp_shr <- 0.002 # def = 0.002
+cfg$gms$s13_max_gdp_shr <- 1 # def = 1
```
I changed this back to 0.002

flohump commented on config/default.cfg:
```diff
@@ -1390,6 +1391,9 @@ cfg$gms$material <- "exo_flexreg_apr16"
 cfg$gms$livestock <- "fbask_jan16"                  # def = fbask_jan16
 
+# * Year until the pasture management factor is fixed to 1
+cfg$gms$s70_past_mngmnt_factor_fix <- "2005"	# def = 2005
```
config added cfg$gms$s70_past_mngmnt_factor_fix with default 2005 (previous default was 2010). The previous setting caused a strong spike in CO2 emissions from pasture expansion in SSA. With 2005, this can be avoided.

flohump commented on core/macros.gms:
```diff
@@ -88,3 +88,16 @@ $macro m_sigmoid_interpol(input,start_year,target_year,start_value,target_value)
     input(t_all)$(m_year(t_all) <= start_year) = start_value; \
     input(t_all)$(m_year(t_all) >= target_year) = target_value;
 
+* macro for simple carbon stocks
```
Carbon stocks for cropland, pasture and urban land are simple to calculate while carbons stocks with age-classes are more difficult.

flohump commented on modules/56_ghg_policy/price_jan20/equations.gms:
```diff
@@ -13,80 +13,97 @@
-                     sum((ct,cell(i2,j2)), p56_emis_policy(ct,i2,pollutants,emis_cell_yr56) *
-                     im_pollutant_prices(ct,i2,pollutants)));
+                     v56_emis_cell_pricing(j2,emis_cell_yr56,pollutants) *
+                     sum((ct,cell(i2,j2)), im_pollutant_prices(ct,i2,pollutants,emis_cell_yr56)));
```
Well, I think you implemented this a couple of years ago. But you are right, it can be done on the regional level. I revised the code accordingly.

flohump commented on config/default.cfg:
```diff
@@ -1198,19 +1201,17 @@ cfg$gms$c56_emis_policy <- "redd+natveg_nosoil" 		# def = redd+natveg_nosoil
+# * 	actual: CO2 emissions for pricing are based on the difference of actual carbon stocks between time steps
+# * 	actualNoAcEst: CO2 emissions for pricing are based on actual carbon stocks but 
+# *			without newly established forest and non-forest areas. Note that re/afforestation is covered by a separate mechanism.
+# * 	previousLandPattern: CO2 emissions for pricing are based on carbon stocks with previous land patterns and current carbon density
```
Yes, there is no endogeneity in this option. This option is only added for completeness.

flohump commented on modules/56_ghg_policy/price_jan20/sets.gms:
```diff
@@ -146,16 +149,22 @@ sets
    emis_reg_yr56(emis_source_reg) regional yearly emission sources
    /inorg_fert, man_crop, awms, resid, man_past, som,
-   rice, ent_ferm,  beccs/
+   rice, ent_ferm /
```
Yes. It was unused for a quite a while.

tscheypidi commented on CHANGELOG.md:
```diff
@@ -33,6 +33,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
  - **scripts** Added script to folder projects paper_grassland.R
  - **44_biodiversity** Improved documentation, simplification of equations and flexible options for price on biodiversity loss
  - **modules** Moved interface `vm_carbon_stock` from 52_carbon to 56_ghg_policy
+ - **config** changed default setting for s56_reward_neg_emis from 0 to -Inf
```
I think you completely removed that switch, right?

flohump replied:
Yes. I changed it in the changelog.

tscheypidi commented on modules/13_tc/endo_jan22/input.gms:
```diff
@@ -8,7 +8,7 @@
 scalars
  s13_ignore_tau_historical	ignore historical tau (1) or use it as lower bound (0) (binary) / 1 /
- s13_max_gdp_shr Maximum tech cost as share of regional GDP / 1 /
+ s13_max_gdp_shr Maximum tech cost as share of regional GDP / 0.002 /
```
that means all runs work again with a max share of 0.002?

flohump replied:
Yes!

bodirsky replied:
0.002 may be a bit low in countries where agriculture is 50% of gdp, no?

tscheypidi commented on modules/35_natveg/dynamic_feb21/input.gms:
```diff
@@ -14,8 +14,8 @@ $setglobal c35_forest_damage_end  by2050
-s35_hvarea_secdforest annual secdforest harvest rate for s35_hvarea equals 1 (percent per year) / 0.005 /
-s35_hvarea_primforest annual primforest harvest rate for s35_hvarea equals 1 (percent per year) / 0.0001 /
+s35_hvarea_secdforest annual secdforest harvest rate for s35_hvarea equals 1 (percent per year) / 0 /
+s35_hvarea_primforest annual primforest harvest rate for s35_hvarea equals 1 (percent per year) / 0 /
```
why has this been changed now?

flohump replied:
Default values set by running the model with `scripts/start/default.R`

tscheypidi commented on modules/44_biodiversity/bv_btc_mar21/input.gms:
```diff
@@ -9,7 +9,7 @@ scalars
  s44_start_price		Price for biodiversity stock loss in start year (USD per ha) 	/ 0 /
- s44_target_price		Price for biodiversity stock loss in target year (USD per ha) 	/ 3000 /
+ s44_target_price		Price for biodiversity stock loss in target year (USD per ha) 	/ 0 /
 ;
```
new default value?

flohump replied:
Default values set by running the model with `scripts/start/default.R`

tscheypidi commented on modules/56_ghg_policy/price_jan20/equations.gms:
```diff
@@ -12,65 +12,47 @@
 
 *** Emissions
 
+*' Actual cellular CO2 emissions are calculated based on changes in carbon stocks between timesteps in the interface `vm_carbon_stock`.
```
why are you talking about cellular emissions if the equation is regional?

flohump replied:
I have updated the documentation.

tscheypidi commented on modules/56_ghg_policy/price_jan20/equations.gms:
```diff
@@ -12,65 +12,47 @@
-	vm_emissions_reg(i2,emis_source_cell,pollutants) =e=
-                sum(cell(i2,j2), vm_btm_cell(j2,emis_source_cell,pollutants)
-                  * (1 - sum(ct, im_maccs_mitigation(ct,i2,emis_source_cell,pollutants))));
+*' Actual GHG emissions
```
Please use complete sentences in the documentation

flohump replied:
done.

tscheypidi commented on modules/56_ghg_policy/price_jan20/preloop.gms:
```diff
@@ -9,12 +9,8 @@
-vm_btm_cell.up(j,emis_co2,"co2_c") = Inf;
-***fix to zero be default here. Bounds are released in the respective modules.
-vm_btm_reg.fx(i,emis_source_reg,pollutants) = 0;
+*vm_btm_reg.fx(i,emis_oneoff,pollutants)$(not sameas(pollutants,"co2_c")) = 0;
```
please delete code, if it is not being used.

flohump replied:
done.

tscheypidi commented on scripts/start/extra/carbon_pricing.R:
```diff
@@ -46,105 +48,105 @@ for(ssp in c("SSP2")) {
+  #   cfg$gms$c56_carbon_stock_pricing <- c56_carbon_stock_pricing
+  #   start_run(cfg, codeCheck = TRUE)
+  # }
 }
```
please delete code if you being used.

flohump replied:
done

bodirsky commented on config/default.cfg:
```diff
@@ -275,7 +275,7 @@ cfg$gms$c13_tccost <- "medium"		# def = medium
 
 # * Maximum regional tech cost expressed as share of reginal GDP
-cfg$gms$s13_max_gdp_shr <- 1 # def = 1
+cfg$gms$s13_max_gdp_shr <- 0.002 # def = 0.002
```
we could also go for 5% if it caused infeasibilties, 100% of gdp just seemd to much...

bodirsky reviewed (approved):
Nice. Reduced code quite a bit.

maybe reconsider the 0.002 as too strict (for low income countries with high ag share).

tscheypidi reviewed (approved):
looks good
